### PR TITLE
RooMCMC class added

### PR DESCRIPTION
**Pull request by odahme**
I added the RooMCMC class into roofitcore. It is used as the RooMinuit class except that it is using a Monte Carlo Markov Chain as a minimizer. I also added a Tutorial to the roofit section where a basic comparison with Minuit can be performed.

This is my first contribution to ROOT so I except that some thing will not be right.
But I compiled ROOT on my computer exactly with the changes I am committing here and everything went fine.

### CODE REVIEW

File: roofit/roofitcore/inc/RooMCMC.h [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+class RooAbsArg ;
+class TVirtualFitter ;
+class TH2D ;
+class TH1F ;
+class RooPlot ;
+class TFile;
+class TMultiGraph;
+class TGraph;
+
+class RooMCMC : public TObject {
```
Comment (pcanal) [inline roofit/roofitcore/inc/RooMCMC.h]: Would it make sense to have a more verbose name (for example be more explicit about " Markov Chain")?

Comment (odahme) [reply roofit/roofitcore/inc/RooMCMC.h]: How about RooMarkov or RooMarkovChain or RooMCMarkovChain?

And sorry for the dumb questions:
What exactly is the PR announcement?
do you mean comment each function int the header file or the cxx file?

Comment (pcanal) [reply roofit/roofitcore/inc/RooMCMC.h]: > What exactly is the PR announcement?

The first 'post' in this conversation :)

Comment (odahme) [reply roofit/roofitcore/inc/RooMCMC.h]: Ah ok I understand I will add it as the documentation of the class right away

Comment (pcanal) [reply roofit/roofitcore/inc/RooMCMC.h]: > How about RooMarkov or RooMarkovChain or RooMCMarkovChain?

I would pick RooMCMarkovChain

**Comment by couet** [comment]
Can you please modify your code to follow the documentation guide lines described here https://root.cern.ch/how/formatting-comments-doxygen ?

File: roofit/roofitcore/inc/RooMCMC.h (hunk 1 above)
Comment (odahme) [reply roofit/roofitcore/inc/RooMCMC.h]: good name is changed. Does anybody else has suggestions for names?

**Comment by odahme** [comment]
I did all the changes you guys requested. What will be the next step? Or are we just waiting for the Jenkins Ci build check to be finished? If yes how long does that usually take?

**Comment by amadio** [comment]
@odahme Jenkins takes quite a bit (maybe a few hours since the nightly builds are running and the system is overloaded at the moment). If the requested changes were addressed, @couet and @pcanal need to approve the changes prior to merging too. I would also like @lmoneta to give his explicit approval, since he is the code owner. So, on your side, you don't have to do anything for the moment. Thanks for the contribution and for addressing everyone's comments!

**Comment by amadio** [comment]
@pcanal @couet @lmoneta Please review this PR again and remove the request for changes if the current version has addressed your comments. Thanks!

**Comment by amadio** [comment]
Unfortunately, this had to be reverted, as it failed the included test in many of the nightly builds.

**Comment by odahme** [comment]
@amadio    that is too bad. What exactly failed? and do you have any idea why it failed?

**Comment by amadio** [comment]
@odahme Here is a sample of the error output (that are many more of these lines in the full output):
```
Processing /Volumes/MacintoshHD2/ec/build/workspace/root-nightly-master-cxx14root7/BUILDTYPE/Release/COMPILER/native/LABEL/mac1012/root/tutorials/roofit/rf611_RooMCMarkovChainexample.C...

[NON-XML-CHAR-0x1B][1mRooFit v3.60 -- Developed by Wouter Verkerke and David Kirkby[NON-XML-CHAR-0x1B][0m 
                Copyright (C) 2000-2013 NIKHEF, University of California & Stanford University
                All rights reserved, please read http://roofit.sourceforge.net/license.txt

[#0] ERROR:Eval -- RooAbsReal::logEvalError(g1) evaluation error, 
 origin       : RooGaussian::g1[ x=x mean=mean1 sigma=sigma1 ]
 message      : p.d.f normalization integral is zero or negative
 server values: x=x=-2.65215, mean=mean1=3.74943, sigma=sigma1=0
[#0] ERROR:Eval -- RooAbsReal::logEvalError(g1) evaluation error, 
 origin       : RooGaussian::g1[ x=x mean=mean1 sigma=sigma1 ]
 message      : p.d.f normalization integral is zero or negative
 server values: x=x=-2.04508, mean=mean1=3.74943, sigma=sigma1=0
[#0] ERROR:Eval -- RooAbsReal::logEvalError(g1) evaluation error, 
 origin       : RooGaussian::g1[ x=x mean=mean1 sigma=sigma1 ]
 message      : p.d.f normalization integral is zero or negative
 server values: x=x=-2.08508, mean=mean1=3.74943, sigma=sigma1=0
[#0] ERROR:Eval -- RooAbsReal::logEvalError(g1) evaluation error, 
 origin       : RooGaussian::g1[ x=x mean=mean1 sigma=sigma1 ]
 message      : p.d.f normalization integral is zero or negative
 server values: x=x=-2.11206, mean=mean1=3.74943, sigma=sigma1=0
[#0] ERROR:Eval -- RooAbsReal::logEvalError(g1) evaluation error, 
 origin       : RooGaussian::g1[ x=x mean=mean1 sigma=sigma1 ]
 message      : p.d.f normalization integral is zero or negative
 server values: x=x=-2.61614, mean=mean1=3.74943, sigma=sigma1=0
[#0] ERROR:Eval -- RooAbsReal::logEvalError(g1) evaluation error, 
 origin       : RooGaussian::g1[ x=x mean=mean1 sigma=sigma1 ]
 message      : p.d.f normalization integral is zero or negative
 server values: x=x=4.19165, mean=mean1=3.74943, sigma=sigma1=0
[#0] ERROR:Eval -- RooAbsReal::logEvalError(g1) evaluation error, 
 origin       : RooGaussian::g1[ x=x mean=mean1 sigma=sigma1 ]
 message      : p.d.f normalization integral is zero or negative
 server values: x=x=-2.07341, mean=mean1=3.74943, sigma=sigma1=0
[#0] ERROR:Eval -- RooAbsReal::logEvalError(g1) evaluation error, 
 origin       : RooGaussian::g1[ x=x mean=mean1 sigma=sigma1 ]
 message      : p.d.f normalization integral is zero or negative
 server values: x=x=3.42638, mean=mean1=3.74943, sigma=sigma1=0
[#0] ERROR:Eval -- RooAbsReal::logEvalError(g1) evaluation error, 
 origin       : RooGaussian::g1[ x=x mean=mean1 sigma=sigma1 ]
 message      : p.d.f normalization integral is zero or negative
 server values: x=x=2.6117, mean=mean1=3.74943, sigma=sigma1=0
[#0] ERROR:Eval -- RooAbsReal::logEvalError(g1) evaluation error, 
 origin       : RooGaussian::g1[ x=x mean=mean1 sigma=sigma1 ]
 message      : p.d.f normalization integral is zero or negative
 server values: x=x=-3.10404, mean=mean1=3.74943, sigma=sigma1=0
[#0] ERROR:Eval -- RooAbsReal::logEvalError(g1) evaluation error, 
 origin       : RooGaussian::g1[ x=x mean=mean1 sigma=sigma1 ]
 message      : p.d.f normalization integral is zero or negative
 server values: x=x=-2.5732, mean=mean1=3.74943, sigma=sigma1=0
```

**Comment by odahme** [comment]
@amadio  ah yes I remember those. look at line 242 of the cxx file. If the calculation of negative log likelihood failes, because the walk got out of bounds it sets the value very high so the algorithm turns away from this point. Minuit needed this error but in a MCMC minimizer it does not matter if it gets out of bounds. I forgot about that problem, the error messages are unnecessary. Thats why they are turned of at line 243. Sorry just turning of the error message was not a good idea did, beginners error.

Do you have a better idea how to solve that problem?

**Comment by amadio** [comment]
@odahme No, sorry, I'm not sure how to fix the problem, that's why I decided to revert the changes. Please discuss this with @lmoneta, as he can help you with that. Turning off the error messages is not a solution. Cheers,

**Pull request merged**